apply/local/applier.go

```diff
@@ -0,0 +1,333 @@
+
+	if err != nil {
+		return err
+	}
+	applier.Runtime, err = runtime.NewDefaultRuntime(applier.ClusterDesired)
+	if err != nil {
+		return fmt.Errorf("failed to init runtime, %v", err)
+	}
+	return nil
+}
+
+func (c *Applier) Delete() (err error) {
+	t := metav1.Now()
+	c.ClusterDesired.DeletionTimestamp = &t
+	return c.Apply()
+}
+
+// Apply different actions between ClusterDesired and ClusterCurrent.
+func (c *Applier) Apply() (err error) {
+	//delete cluster
+	if c.ClusterDesired.DeletionTimestamp != nil {
+		return c.deleteCluster()
+	}
+
+	err = c.fillClusterCurrent()
+	if err != nil {
+		return err
+	}
+
+	// first time to init cluster
+	if c.ClusterCurrent == nil {
+		return c.initCluster()
+	}
+
+	// change cluster, such as upgrade,scale,install app on an existed cluster.
+	// infra change: scaleCluster: prepare k8s env,apply node and master,apply guest
+	// application change:upgradeCluster; install app:
+	// infra change and application change could happen at the same time.
+	c.MastersToJoin, c.MastersToDelete = utils.GetDiffHosts(c.ClusterCurrent.Spec.Masters, c.ClusterDesired.Spec.Masters)
+	c.NodesToJoin, c.NodesToDelete = utils.GetDiffHosts(c.ClusterCurrent.Spec.Nodes, c.ClusterDesired.Spec.Nodes)
+
+	if c.ClusterDesired.Spec.Image != c.ClusterCurrent.Spec.Image {
+		logger.Info("current image is : %s and desired image is : %s , so upgrade your cluster",
+			c.ClusterCurrent.Spec.Image, c.ClusterDesired.Spec.Image)
+		return c.upgradeCluster()
+	}
+
+	return c.scaleCluster()
+}
+
+func (c *Applier) fillClusterCurrent() error {
+	currentCluster, err := GetCurrentCluster(c.Client)
+	if err != nil {
+		return errors.Wrap(err, "get current cluster failed")
+	}
+	if currentCluster != nil {
+		c.ClusterCurrent = c.ClusterDesired.DeepCopy()
+		c.ClusterCurrent.Spec.Masters = currentCluster.Spec.Masters
+		c.ClusterCurrent.Spec.Nodes = currentCluster.Spec.Nodes
+		//upgrade
+		clusterFilePath := common.GetClusterWorkClusterfile(c.ClusterDesired.Name)
+		if utils.IsFileExist(clusterFilePath) {
+			clusterBeforeUpgrade, err := utils.GetClusterFromFile(clusterFilePath)
+			if err != nil {
+				return err
+			}
+			if clusterBeforeUpgrade.Spec.Image != c.ClusterDesired.Spec.Image {
+				c.ClusterCurrent.Spec.Image = clusterBeforeUpgrade.Spec.Image
+			}
+		}
+	}
+	// all commands need to do
+	err = utils.SaveClusterfile(c.ClusterDesired)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *Applier) upgradeCluster() error {
+	var todoList []ActionName
+	todoList = append(todoList, PullIfNotExist)
+	todoList = append(todoList, MountImage)
+	todoList = append(todoList, MountRootfs)
+	todoList = append(todoList, Upgrade)
+	if len(c.MastersToJoin) > 0 || len(c.MastersToDelete) > 0 {
+		todoList = append(todoList, ApplyMasters)
+	}
+	if len(c.NodesToJoin) > 0 || len(c.NodesToDelete) > 0 {
+		todoList = append(todoList, ApplyNodes)
+	}
+	todoList = append(todoList, Guest)
```

Comment:
I think it's better to delete L252-L257,because these codes complete the scale task

Reply:
But in terms of this PR,func Apply() return at L211,so if delete these lines,it won't work when upgrade and scale at the same time.

Reply:
@shy-xu, yes, i didn’t modify the upgrade logic. those code looks like to make sure it wont failed when upgrade and scale at the same time.